gen/map.go

```diff
@@ -281,3 +281,147 @@ func (m *mapGenerator) equalsUnhashable(g Generator, spec *compile.MapSpec) (str
 
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
+
+func (m *mapGenerator) zapMarshaler(
+	g Generator,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	name := zapperName(g, root)
+	switch compile.RootTypeSpec(root.KeySpec).(type) {
+	case *compile.StringSpec:
+		return m.zapStringKeyMarshaler(g, name, root, fieldValue)
+	default:
+		return m.zapNonstringKeyMarshaler(g, name, root, fieldValue)
+	}
+}
+
+func (m *mapGenerator) zapStringKeyMarshaler(
+	g Generator,
+	name string,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	if err := g.EnsureDeclared(
+		`
+			<$zapcore := import "go.uber.org/zap/zapcore">
+
+			type <.Name> <typeReference .Type>
+			<$m := newVar "m">
+			<$k := newVar "k">
+			<$v := newVar "v">
+			<$enc := newVar "enc">
+			// MarshalLogObject implements zapcore.ObjectMarshaler, enabling
+			// fast logging of <.Name>.
+			func (<$m> <.Name>) MarshalLogObject(<$enc> <$zapcore>.ObjectEncoder) error {
+				for <$k>, <$v> := range <$m> {
+					<zapEncodeBegin .Type.ValueSpec ->
+						<$enc>.Add<zapEncoder .Type.ValueSpec>((string)(<$k>), <zapMarshaler .Type.ValueSpec $v>)
+					<- zapEncodeEnd .Type.ValueSpec>
+				}
+				return nil
+			}
+			`, struct {
+			Name string
+			Type *compile.MapSpec
+		}{
+			Name: name,
+			Type: root,
+		},
+	); err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("(%v)(%v)", name, fieldValue), nil
+}
+
+func (m *mapGenerator) zapNonstringKeyMarshaler(
+	g Generator,
+	name string,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	if err := g.EnsureDeclared(
+		`
+			<$zapcore := import "go.uber.org/zap/zapcore">
+
+			type <.Name> <typeReference .Type>
+			<$m := newVar "m">
+			<$k := newVar "k">
+			<$v := newVar "v">
+			<$i := newVar "i">
+			<$enc := newVar "enc">
+			// MarshalLogArray implements zapcore.ArrayMarshaler, enabling
+			// fast logging of <.Name>.
+			func (<$m> <.Name>) MarshalLogArray(<$enc> <$zapcore>.ArrayEncoder) error {
+				<- if isHashable .Type.KeySpec ->
+					for <$k>, <$v> := range <$m> {
+				<else ->
+					for _, <$i> := range <$m> {
+						<$k> := <$i>.Key
+						<$v> := <$i>.Value
+				<end ->
+					if err := <$enc>.AppendObject(<zapMapItemMarshaler .Type.KeySpec $k .Type.ValueSpec $v>); err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+			`, struct {
+			Name string
+			Type *compile.MapSpec
+		}{
+			Name: name,
+			Type: root,
+		},
+		TemplateFunc("zapMapItemMarshaler", curryGenerator(m.zapMapItemMarshaler, g)),
```

Comment:
Looking at TemplateFunc docs, you don't need to `curryGenerator`. That's done
automatically if the provided function accepts a Generator as its first
argument.